### PR TITLE
WIP: Async Recipe creation

### PR DESCRIPTION
**Pull request by jbusecke**
Scaling the recipe up to 100s of datasets in #11 is grinding the registrar to a halt at the moment. 

I have experimented locally with an async version that gives me about 10x speedup. 

Lets see if that works here too.

### CODE REVIEW

**Comment by jbusecke** [comment]
Holy moly, this worked! AHAHAHAHAHAHA

File: feedstock/recipe.py [hunk 1]
```diff
@@ -354,114 +311,123 @@ def _check_single_element_list(lst):
...
+        for iid in iids:
+            tasks.append(asyncio.ensure_future(iid_request(session, iid, node_dict.values())))
+
+        raw_input = await asyncio.gather(*tasks)
+        recipe = {iid:recipe_from_urls(urls, kwargs) for iid, (urls, kwargs) in zip(iids, raw_input) if urls is not None}
+        return recipe
+
+# If you want to debug this in a jupyter notebook you need to uncomment the code below and instead import main and then do `await main()` (see 
+recipes = asyncio.run(main())
+#Final TODO: How do I incorporate this into pangeo forge?
```
Comment (cisaacstern) [inline feedstock/recipe.py:433]: > #Final TODO: How do I incorporate this into pangeo forge? 

@jbusecke what does this comment mean? Anything I can do to help?

Comment (jbusecke) [reply feedstock/recipe.py:433]: Haha no I was expecting this to actually fail, but it didnt!

**Comment by jbusecke** [comment]
There still seems to be a bug that puts recipes in there which have failed during the creation? Or maybe its just flakyness on the ESGF end. will investigate in a bit!
Or we merge this now and see which recipes fail?

**Comment by cisaacstern** [comment]
> Or we merge this now and see which recipes fail?

You're the lead maintainer here now, so whatever you think is best :)

**Comment by jbusecke** [comment]
A couple of things that I just wanted to note down here: 
- I am seeing a bunch of 503s (e.g. [here](https://github.com/pangeo-forge/registrar/runs/7047017180?check_suite_focus=true#step:2:846) that seem flaky since they would fail/pass within different runs during my local testing). It would be very nice if we could implement a clever retry strategy here.

File: feedstock/recipe.py [hunk 2]
```diff
@@ -1,64 +1,41 @@
...
-
-import requests
 import pandas as pd
 from typing import List, Dict, Tuple
 from pangeo_forge_recipes.patterns import pattern_from_file_sequence
 from pangeo_forge_recipes.recipes import XarrayZarrRecipe
 
+import aiohttp
+import asyncio
+import time
```
Comment (jbusecke) [inline feedstock/recipe.py:8]:
```suggestion
```

**Comment by cisaacstern** [comment]
> Unable to open dataset with dataset_public_url = 'https://ncsa.osn.xsede.org/Pangeo/pangeo-forge-test/prod/recipe-run-644/pangeo-forge/cmip6-feedstock/CMIP6.DAMIP.MOHC.HadGEM3-GC31-LL.hist-aer.r4i1p1f3.Amon.pr.gn.v20190814.zarr'

This is not a common error. It suggests a recipe was successfully run but the bot could not open the resulting dataset. Have you tried opening that url manually? Perhaps it requires some extra kwargs or something?

**Comment by jbusecke** [comment]
I just tried that and got `KeyError: 'Zarr object is missing the attribute `_ARRAY_DIMENSIONS`, which is required for xarray to determine variable dimensions.'`

**Comment by cisaacstern** [comment]
That's strange. I'm not sure what that means.

**Comment by jbusecke** [comment]
Neither do I...should I just try to rerun this and see if it was a fluke?

**Pull request merged**